src/libraries/Common/src/Interop/Browser/Interop.Runtime.NativeAOT.cs

```diff
@@ -0,0 +1,77 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.JavaScript;
+
+internal static partial class Interop
+{
+    // WARNING: until https://github.com/dotnet/runtime/issues/37955 is fixed
+    // make sure that the native side always sets the out parameters
+    // otherwise out parameters could stay un-initialized, when the method is used in inlined context
+    internal static unsafe partial class Runtime
+    {
+#pragma warning disable SYSLIB1054
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal static extern void ReleaseCSOwnedObject(IntPtr jsHandle);
+        [DllImport("*", EntryPoint = "mono_wasm_bind_js_function")]
+        public static extern unsafe void BindJSFunction(string function_name, int function_name_length, string module_name, int module_name_length, void* signature, out IntPtr bound_function_js_handle, out int is_exception);
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        public static extern void InvokeJSFunction(IntPtr bound_function_js_handle, void* data);
+        [DllImport("*", EntryPoint = "mono_wasm_invoke_import")]
+        public static extern unsafe void InvokeImport(IntPtr fn_handle, void* data);
+        [DllImport("*", EntryPoint = "mono_wasm_bind_cs_function")]
+        public static extern unsafe void BindCSFunction(string fully_qualified_name, int fully_qualified_name_length, int signature_hash, void* signature, out int is_exception);
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        public static extern void ResolveOrRejectPromise(void* data);
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        public static extern IntPtr RegisterGCRoot(IntPtr start, int bytesSize, IntPtr name);
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        public static extern void DeregisterGCRoot(IntPtr handle);
+#pragma warning restore SYSLIB1054
+
+        public static unsafe void BindJSFunction(string function_name, string module_name, void* signature, out IntPtr bound_function_js_handle, out int is_exception, out object result)
+        {
+            BindJSFunction(function_name, function_name.Length, module_name, module_name.Length, signature, out bound_function_js_handle, out is_exception);
+            if (is_exception != 0)
+                result = "Runtime.BindJSFunction failed";
+            else
+                result = "";
```

Comment:
The original Mono functions were internal calls and would run in cooperative GC mode (i. e. as if they were managed code). `DllImport` is PInvoke, which puts the code into preemptive mode. I wonder if this will cause problems.

(Code in preemptive mode moslty cannot touch managed state).

Reply:
what's the plan for running GC for NAOT on browser ? What would trigger it and what would schedule finalizers ?

Reply:
> what's the plan for running GC for NAOT on browser ? What would trigger it and what would schedule finalizers ?

Same as on other platforms - you GC whenever GC decides its time to GC. Practically, this will happens on allocations, or explicit `GC.Collect`s.

Finalization is a little bit NYI right now (#2240), but it could follow the same plan (nice because WASI would _have_ to do that), or be hooked into the event loop.

Reply:
I guess we need to make decision about yielding to browser event loop. 
Big part of JS interop is moot without it (promises, network events, UI).

Reply:
Well, I would expect the yielding to happen naturally in async code, as you hit `await`s and such. GC shouldn't really play much into it? We would of course have the Browser-specific thread pool implementation.